metals/src/main/scala/scala/meta/internal/metals/BSPErrorHandler.scala

```diff
@@ -0,0 +1,172 @@
+        Future.successful(languageClient.showMessage(params))
+      case Left(params) =>
+        languageClient.showMessageRequest(params).asScala.map {
+          case msg if msg == BSPErrorHandler.goToLogs =>
+            val errorMsgStartLine =
+              message.linesIterator.headOption
+                .flatMap(findLine(_))
+                .getOrElse(0)
+            gotoLogs(errorMsgStartLine)
+          case _ => Future.successful(())
+        }
+    }
+  }
+
+  private def findLine(line: String): Option[Int] =
+    try {
+      val (_, lineNumber) =
+        Files
+          .newBufferedReader(logsPath.toNIO)
+          .lines()
+          .asScala
+          .zipWithIndex
+          .fold(("", 0)) {
+            case ((lineBefore, matchingLine), (currentLine, currentIndex)) =>
+              if (
+                lineBefore.contains(BSPErrorHandler.errorInBsp) &&
+                currentLine.contains(line)
+              ) (currentLine, currentIndex)
+              else (currentLine, matchingLine)
+          }
+      Some(lineNumber)
+    } catch {
+      case NonFatal(_) => None
+    }
+
+  private def gotoLogs(line: Int) = {
+    val pos = new l.Position(line, 0)
+    val location = new l.Location(
+      logsPath.toURI.toString(),
+      new l.Range(pos, pos),
+    )
+    languageClient.metalsExecuteClientCommand(
+      ClientCommands.GotoLocation.toExecuteCommandParams(
+        ClientCommands.WindowLocation(
+          location.getUri(),
+          location.getRange(),
+        )
+      )
+    )
+  }
+}
+
+object BSPErrorHandler {
+  def makeShowMessage(
+      message: String
+  ): Either[l.ShowMessageRequestParams, l.MessageParams] = {
+    if (message.length() <= MESSAGE_MAX_LENGTH) {
+      val messageParams = new l.MessageParams(
+        l.MessageType.Error,
+        s"""|$errorHeader:
+            |$message""".stripMargin,
+      )
+      Right(messageParams)
+    } else {
+      val params = new l.ShowMessageRequestParams()
+      params.setMessage(
+        s"""|$errorHeader
+            |${message.take(MESSAGE_MAX_LENGTH)}...
+            |$gotoLogsToSeeFull""".stripMargin
+      )
+      params.setType(l.MessageType.Error)
+      params.setActions(List(goToLogs).asJava)
+      Left(params)
+    }
+  }
+
+  private val errorHeader = "Encountered an error in the build server:"
+  private val goToLogs = new l.MessageActionItem("Go to logs.")
+  private val gotoLogsToSeeFull = "Go to logs to see the full error"
+  private val errorInBsp = "Error in BSP:"
+
+  private val MESSAGE_MAX_LENGTH = 150
+
+}
+
+class CurrentAndPreviousError(
+    previousError: Option[String] = None,
+    currentError: String = "",
+) {
+  lazy val error: Option[String] = {
+    if (currentError.isEmpty()) None
+    else {
+      val deduplicatedError = {
+        currentError.linesIterator.headOption
+          .map { firstLine =>
+            currentError.linesIterator
+              .foldLeft(List.empty[List[String]]) {
```

Comment:
Is this fold needed? Doesn't this just remove the first line?

If not could you explain a bit more deduplicaing the error (maybe in comment?)

Reply:
It divides into parts all starting with the first line. So if you have the same text twice or more it gets deduplicated later. But I'll try to refactor since it's clearly not very readable.